hazelcast-jet-core/src/main/java/com/hazelcast/jet/windowing/SlidingWindowP.java

```diff
@@ -100,34 +102,38 @@ protected boolean tryProcessPunc0(@Nonnull Punctuation punc) {
     }
 
     private Traverser<Object> slidingWindowTraverser(Punctuation punc) {
-        Traverser<Object> traverser = Traversers.traverseStream(
-                range(nextFrameSeqToEmit, nextFrameSeqToEmit = wDef.higherFrameSeq(punc.seq()), wDef.frameLength())
-                        .mapToObj(frameSeq ->
-                                computeWindow(frameSeq, seqToKeyToFrame.remove(frameSeq - wDef.windowLength()))
-                                        .entrySet().stream()
-                                        .map(e -> new Frame<>(frameSeq, e.getKey(), finishF.apply(e.getValue()))))
-                        .flatMap(identity()));
-
-        traverser = traverser.append(punc);
-        return traverser;
+        long rangeStart = nextFrameSeqToEmit;
+        nextFrameSeqToEmit = wDef.higherFrameSeq(punc.seq());
+        return Traversers.traverseStream(range(rangeStart, nextFrameSeqToEmit, wDef.frameLength()).boxed())
+                .flatMap(frameSeq -> Traversers.traverseIterable(computeWindow(frameSeq).entrySet())
+                        .map(e -> (Object) new Frame<>(frameSeq, e.getKey(), finishF.apply(e.getValue())))
+                        .onFirstNull(() -> completeWindow(frameSeq)))
+                .append(punc);
     }
 
-    private Map<K, F> computeWindow(long frameSeq, Map<K, F> evictedFrame) {
+    private void completeWindow(long frameSeq) {
+        Map<K, F> evictedFrame = seqToKeyToFrame.remove(frameSeq - wDef.windowLength() + wDef.frameLength());
         if (deductF != null) {
-            // add leading-edge frame
```

Comment:
Marko, do you believe, that this is so much more complex as to justify to afford 50% more memory in special, but real case and a bit more memory in all cases? This is just inconsistent with what you did in `WindowDefinition.higherFrameSeq()` for a case, that will _never_ happen. It could be just one-liner instead of 4 lines, and you argued "it's good to be correct even in edge cases". Why not this time?

Btw, we should special-case tumbling windows: we'll deduct all keys to zero or unnecessarily copy the frame, if there is no `deductF`. But will this be too much complexity? I don't dare to answer.

And we could also avoid the `onFirstNull` if you don't like it, in `tryProcessPunc()` create a traverser just for single frameSeq. We'll then have a chance to clean up after `emitFromTraverser` returns `true`. This is how I originally written it, before you converted it to use `flatMap`. `Stream` or `Traverser` doesn't make everything less complex, just because it's hip.

And the punctuation N before or after frameN: this is another issue I intended to raise, after this is merged. It should be before, and after frameN, there should be punct(N+1) saying, that frameN is complete.